File: learn-haskell/Advent_of_Code/Day1/scala/src/main/scala/Day1.scala
import scala.annotation.tailrec

object Day1 extends App {
  val s = scala.io.Source.fromFile(args(0)).getLines.toSeq.map(_.toLong)
  val r = pair(s).filter(_.sum == 2020).map(_.product).head
  println(r)

  def _pair(x: Long,xs: Seq[Long]): Seq[Seq[Long]] = {
    @tailrec
    def _pair_tailrec(x: Long, xs: Seq[Long], acc: Seq[Seq[Long]]): Seq[Seq[Long]] = {
      (x, xs) match {
        case (_, Nil) => Nil
        case (n, x :: Nil) => acc ++ Seq(Seq(n, x))
        case (n, x :: xs) => _pair_tailrec(n,xs,acc ++ Seq(Seq(n,x)))
      }

    }
    _pair_tailrec(x, xs, Seq())
  }

  def pair(a: Seq[Long]): Seq[Seq[Long]] = {
    @tailrec
    def pair_tailrec(a: Seq[Long], acc: Seq[Seq[Long]]): Seq[Seq[Long]] = {
      a match {
        case Nil => acc
        case x :: xs => pair_tailrec(xs, acc ++ _pair(x, xs))
      }
    }
    pair_tailrec(a, Seq())
  }
}

File: learn-haskell/Advent_of_Code/Day1/day1.hs
import System.Environment
import System.IO
import System.IO.Error
import Control.Exception

main = catch toTry handler

toTry :: IO () 
toTry = do
    (filename:_) <- getArgs
    contents <- readFile filename
    print . day1 . fmap (read::String->Int) . lines $ contents

handler :: IOError -> IO ()
handler e 
  | isDoesNotExistError e = putStrLn "The file does not exist"
  | otherwise = ioError e

--
--
--
day1 :: (Num a, Eq a) => [a] -> a
day1 xs = head . fmap (product) . filter (\x -> (sum x) == 2020) . pair $ xs

pair :: (Num a) => [a] -> [[a]]
pair [] = []
pair (x:xs) = pair' x xs ++ pair xs
  where 
    pair' :: Num a => a -> [a] -> [[a]]
    pair' _ [] = []
    pair' n [x] = [[n,x]]
    pair' n (x:xs) = [[n,x]] ++ pair' n xs



